openmeter/billing/adapter/profile.go

```diff
@@ -0,0 +1,319 @@
+		SetNamespace(input.Namespace).
+		SetNillableTimezone(input.WorkflowConfig.Timezone).
+		SetCollectionAlignment(input.WorkflowConfig.Collection.Alignment).
+		SetItemCollectionPeriodSeconds(int64(input.WorkflowConfig.Collection.ItemCollectionPeriod / time.Second)).
+		SetInvoiceAutoAdvance(input.WorkflowConfig.Invoicing.AutoAdvance).
+		SetInvoiceDraftPeriodSeconds(int64(input.WorkflowConfig.Invoicing.DraftPeriod / time.Second)).
+		SetInvoiceDueAfterSeconds(int64(input.WorkflowConfig.Invoicing.DueAfter / time.Second)).
+		SetInvoiceItemResolution(input.WorkflowConfig.Invoicing.ItemResolution).
+		SetInvoiceItemPerSubject(input.WorkflowConfig.Invoicing.ItemPerSubject).
+		SetInvoiceCollectionMethod(input.WorkflowConfig.Payment.CollectionMethod).
+		Save(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	dbProfile, err := c.BillingProfile.Create().
+		SetNamespace(input.Namespace).
+		SetKey(input.Key).
+		SetDefault(input.Default).
+		SetTaxProvider(input.TaxConfiguration.Type).
+		SetInvoicingProvider(input.InvoicingConfiguration.Type).
+		SetPaymentProvider(input.PaymentConfiguration.Type).
+		SetSupplierName(input.Supplier.Name).
+		SetSupplierAddressCountry(*input.Supplier.Address.Country). // Validation is done at service level
+		SetNillableSupplierAddressState(input.Supplier.Address.State).
+		SetNillableSupplierAddressCity(input.Supplier.Address.City).
+		SetNillableSupplierAddressPostalCode(input.Supplier.Address.PostalCode).
+		SetNillableSupplierAddressLine1(input.Supplier.Address.Line1).
+		SetNillableSupplierAddressLine2(input.Supplier.Address.Line2).
+		SetNillableSupplierAddressPhoneNumber(input.Supplier.Address.PhoneNumber).
+		SetWorkflowConfig(dbWorkflowConfig).
+		Save(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Hack: we need to add the edges back
+	dbProfile.Edges.WorkflowConfig = dbWorkflowConfig
+
+	return mapProfileFromDB(dbProfile), nil
+}
+
+func (a adapter) GetProfileByKeyOrID(ctx context.Context, input billing.GetProfileByKeyOrIDInput) (*billing.Profile, error) {
+	if err := input.Validate(); err != nil {
+		return nil, err
+	}
+
+	dbProfiles, err := a.client().BillingProfile.Query().
+		Where(billingprofile.Namespace(input.Namespace)).
+		Where(billingprofile.Or(
+			billingprofile.And(billingprofile.Key(input.IDOrKey), billingprofile.DeletedAtIsNil()),
+			billingprofile.ID(input.IDOrKey),
+		)).
+		WithWorkflowConfig().All(ctx)
+	if err != nil {
+		if db.IsNotFound(err) {
+			return nil, nil
+		}
+
+		return nil, err
+	}
+
+	for _, dbProfile := range dbProfiles {
+		if dbProfile.Key == input.IDOrKey {
+			return mapProfileFromDB(dbProfile), nil
+		}
+	}
+
+	for _, dbProfile := range dbProfiles {
+		if dbProfile.ID == input.IDOrKey {
+			return mapProfileFromDB(dbProfile), nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (a adapter) GetProfileByID(ctx context.Context, input billing.GetProfileByIDAdapterInput) (*billing.Profile, error) {
+	if err := input.Validate(); err != nil {
```

Comment:
why not call `GetProfileByKeyOrID` from it?

Reply:
Because:
- KeyOrID is only used if the service don't know which one to fetch (e.g. when it's coming from the frontend)
- We call ID explicitly in the code if we want to lookup by ID
- We call the getByKey if we want to look up by Key

I think it's cleaner to limit the KeyOrID's ambiguity as much as possible. The only reason this exists is because we will have KeyOrID queries and I wanted to have a single query running in such cases.

Also I am not huge fan of even the slightest possibility to resolve a ULID formatted key by accident ;-)